packages/core/src/components/forms/useAsyncControllableValue.ts

```diff
@@ -0,0 +1,138 @@
+/* !
+ * (c) Copyright 2023 Palantir Technologies Inc. All rights reserved.
+ */
+
+import * as React from "react";
+
+interface IUseAsyncControllableValueProps<E extends HTMLInputElement | HTMLTextAreaElement> {
+    value?: React.InputHTMLAttributes<E>["value"];
+    onChange?: React.ChangeEventHandler<E>;
+    onCompositionStart?: React.CompositionEventHandler<E>;
+    onCompositionEnd?: React.CompositionEventHandler<E>;
+}
+
+/**
+ * The amount of time (in milliseconds) which the input will wait after a compositionEnd event before
+ * unlocking its state value for external updates via props. See `handleCompositionEnd` for more details.
+ */
+export const ASYNC_CONTROLLABLE_VALUE_COMPOSITION_END_DELAY = 10;
+
+/*
+ * A hook to workaround the following [React bug](https://github.com/facebook/react/issues/3926).
+ * This bug is reproduced when an input receives CompositionEvents
+ * (for example, through IME composition) and has its value prop updated asychronously.
+ * This might happen if a component chooses to do async validation of a value
+ * returned by the input's `onChange` callback.
+ */
+export function useAsyncControllableValue<E extends HTMLInputElement | HTMLTextAreaElement>(
+    props: IUseAsyncControllableValueProps<E>,
+) {
+    const { onCompositionStart, onCompositionEnd, value: propValue, onChange } = props;
+
+    // The source of truth for the input value. This is not updated during IME composition.
+    // It may be updated by a parent component.
+    const [value, setValue] = React.useState(propValue);
+
+    // The latest input value, which updates during IME composition.
+    const [nextValue, setNextValue] = React.useState(propValue);
+
+    // Whether we are in the middle of a composition event.
+    const [isComposing, setIsComposing] = React.useState(false);
+
+    // Whether there is a pending update we are expecting from a parent component.
+    const [hasPendingUpdate, setHasPendingUpdate] = React.useState(false);
+
+    const cancelPendingCompositionEnd = React.useRef<(() => void) | null>(null);
```

Comment:
Does the suggestion below also work? I saw that the original version used `null` instead of `undefined` to be more clear, but I think `useRef`'s typings make `undefined` a bit more ergonomic. Feel free to ignore this suggestion if you disagree.

```suggestion
    const cancelPendingCompositionEnd = React.useRef<() => void>();
```

Reply:
Awesome, thanks!